test/extensions/filters/network/local_ratelimit/local_ratelimit_test.cc

```diff
@@ -0,0 +1,199 @@
+public:
+  struct ActiveFilter {
+    ActiveFilter(const ConfigSharedPtr& config) : filter_(config) {
+      filter_.initializeReadFilterCallbacks(read_filter_callbacks_);
+    }
+
+    NiceMock<Network::MockReadFilterCallbacks> read_filter_callbacks_;
+    Filter filter_;
+  };
+};
+
+// Basic no rate limit case.
+TEST_F(LocalRateLimitFilterTest, NoRateLimit) {
+  setup(R"EOF(
+stat_prefix: local_rate_limit_stats
+token_bucket:
+  max_tokens: 1
+  fill_interval: 0.2s
+)EOF");
+
+  InSequence s;
+  ActiveFilter active_filter(config_);
+  EXPECT_EQ(Network::FilterStatus::Continue, active_filter.filter_.onNewConnection());
+  EXPECT_EQ(0, TestUtility::findCounter(stats_store_,
+                                        "local_rate_limit.local_rate_limit_stats.rate_limit")
+                   ->value());
+}
+
+// Basic rate limit case.
+TEST_F(LocalRateLimitFilterTest, RateLimit) {
+  setup(R"EOF(
+stat_prefix: local_rate_limit_stats
+token_bucket:
+  max_tokens: 1
+  fill_interval: 0.2s
+)EOF");
+
+  // First connection is OK.
+  InSequence s;
+  ActiveFilter active_filter1(config_);
+  EXPECT_EQ(Network::FilterStatus::Continue, active_filter1.filter_.onNewConnection());
+
+  // Second connection should be rate limited.
+  ActiveFilter active_filter2(config_);
+  EXPECT_CALL(active_filter2.read_filter_callbacks_.connection_, close(_));
+  EXPECT_EQ(Network::FilterStatus::StopIteration, active_filter2.filter_.onNewConnection());
+  EXPECT_EQ(1, TestUtility::findCounter(stats_store_,
+                                        "local_rate_limit.local_rate_limit_stats.rate_limit")
+                   ->value());
+
+  // Refill the bucket.
+  EXPECT_CALL(*fill_timer_, enableTimer(std::chrono::milliseconds(200), nullptr));
+  fill_timer_->invokeCallback();
+
+  // Third connection is OK.
+  ActiveFilter active_filter3(config_);
+  EXPECT_EQ(Network::FilterStatus::Continue, active_filter3.filter_.onNewConnection());
+}
+
+// Verify the runtime disable functionality.
+TEST_F(LocalRateLimitFilterTest, RuntimeDisabled) {
+  setup(R"EOF(
+stat_prefix: local_rate_limit_stats
+token_bucket:
+  max_tokens: 1
+  fill_interval: 0.2s
+enabled:
+  default_value: true
+  runtime_key: foo_key
+)EOF");
+
+  // First connection is OK.
+  InSequence s;
+  ActiveFilter active_filter1(config_);
+  EXPECT_CALL(runtime_.snapshot_, getBoolean("foo_key", true)).WillOnce(Return(true));
+  EXPECT_EQ(Network::FilterStatus::Continue, active_filter1.filter_.onNewConnection());
+
+  // Second connection should be rate limited but won't be due to filter disable.
+  ActiveFilter active_filter2(config_);
+  EXPECT_CALL(runtime_.snapshot_, getBoolean("foo_key", true)).WillOnce(Return(false));
+  EXPECT_EQ(Network::FilterStatus::Continue, active_filter2.filter_.onNewConnection());
+  EXPECT_EQ(0, TestUtility::findCounter(stats_store_,
+                                        "local_rate_limit.local_rate_limit_stats.rate_limit")
+                   ->value());
+}
```

Comment:
My only test concern is we don't really have anything that verifies the compare_exchange_weak models.
Would it be fairly straight-forward to have a test with 10-20 tokens and N+1 threads which each try to get a token, then make sure one got rejected and all the others succeeded?  Ditto on testing concurrency onFillTimer

Reply:
I'm going to add something like https://github.com/apache/incubator-pagespeed-mod/blob/master/pagespeed/kernel/thread/thread_synchronizer.h (per @jmarantz, I've also done similar things in previous projects). This will allow for deterministic multi-threaded tests of these edge cases. I will try to get this done today but it might not be until EOW since I am out tomorrow and Wednesday. cc @junr03